### PR TITLE
fix: return the correct default trait collection

### PR DESCRIPTION
**Pull request by alanjhughes**
## Summary:
Currently, when the `_currentColorScheme` is nil the default comes from `[UITraitCollection currentTraitCollection]` which provides the trait collection of the context it was called in. Generally, this will work but in some cases the context can be different and this will return the wrong value also if the `Appearance` property is set in the plist, then initially that value is returned, regardless of if you have overridden it.  Seen as the `userInterfaceStyle` of the window is overridden when the value is changed, then the default should be whatever the current style of the window is and not dependent on the calling context.

<table>
    <tr><th>Before</th><th>After</th></tr>
    <tr>
    <td>
        <video src="https://github.com/facebook/react-native/assets/30924086/3bd2d217-00f2-41d5-9229-05c31da2ecf5"/>
   </td>
   <td>
       <video src="https://github.com/facebook/react-native/assets/30924086/6ed67e4c-dd01-40ff-84fd-0d7ffdf6534c"   />  
    </td>
</tr> 
</table> 

## Changelog:

[IOS] [FIXED] - Fix the default trait collection to always return the value of the window

## Test Plan:

Tested on rn-tester and verified the current behaviour is unaffected and also in our own repo to make sure it addresses the problem. Video provided above

### CODE REVIEW

File: packages/react-native/React/CoreModules/RCTAppearance.mm [hunk 1]
```diff
@@ -36,6 +36,18 @@ void RCTOverrideAppearancePreference(NSString *const colorSchemeOverride)
...
+  if ([NSThread isMainThread]) {
+    return [UIApplication sharedApplication].delegate.window.traitCollection;
+  } else {
+    __block UITraitCollection* traitCollection = nil;
+    dispatch_sync(dispatch_get_main_queue(), ^{
+      traitCollection = [UIApplication sharedApplication].delegate.window.traitCollection;
+    });
+    return traitCollection;
+  }
+}
```
Comment (cipolleschi) [inline packages/react-native/React/CoreModules/RCTAppearance.mm:49]: I think we can simplify this code by using [`RCTExecuteOnMainQueue`](https://github.com/facebook/react-native/blob/9841bd81852d59608fe3566b17831d6d42eb7dcf/packages/react-native/React/Base/RCTUtils.m#L259)

**Review by cipolleschi — CHANGES_REQUESTED** [review]
Hi @alanjhughes, thanks for taking the time for doing this.
I think this can fix multiple issue we had on React Native related to the same root cause, so I'm very keen to land it.
I suggested a change to simplify the code: once updated I'll import it.... Hoping that our internal tests will actually pass. If that won't be the case, we would have to figure out a way to solve this... but hopefully there would not be the need!

File: packages/react-native/React/CoreModules/RCTAppearance.mm [hunk 2]
```diff
@@ -36,6 +36,18 @@ void RCTOverrideAppearancePreference(NSString *const colorSchemeOverride)
   return sColorSchemeOverride;
 }
 
+UITraitCollection* getKeyWindowTraitCollection() {
+  if ([NSThread isMainThread]) {
+    return [UIApplication sharedApplication].delegate.window.traitCollection;
+  } else {
+    __block UITraitCollection* traitCollection = nil;
+    dispatch_sync(dispatch_get_main_queue(), ^{
+      traitCollection = [UIApplication sharedApplication].delegate.window.traitCollection;
```
Comment (cipolleschi) [inline packages/react-native/React/CoreModules/RCTAppearance.mm:45]: Please, use the [`RCTSharedApplication`](https://github.com/facebook/react-native/blob/9841bd81852d59608fe3566b17831d6d42eb7dcf/packages/react-native/React/Base/RCTUtils.m#L537)  utility function here

**Comment by alanjhughes** [comment]
Thanks for the review @cipolleschi Sorry for the delay. I've made the changes

**Review by cipolleschi — APPROVED** [review]

**Comment by alanjhughes** [comment]
#39439

**Pull request closed without merging**